vortex-array/src/compute/filter.rs

```diff
@@ -66,9 +77,53 @@ pub fn filter(array: &ArrayData, mask: &FilterMask) -> VortexResult<ArrayData> {
 pub struct FilterMask {
     array: ArrayData,
     true_count: usize,
+    true_range: (usize, usize),
+    selectivity: f64,
+    indices: OnceLock<Vec<usize>>,
+    slices: OnceLock<Vec<(usize, usize)>>,
     buffer: OnceLock<BooleanBuffer>,
 }
 
+/// We implement Clone manually to trigger population of our cached indices or slices.
+/// By making the filter API take FilterMask by value, whenever it gets used multiple times
+/// in a recursive function, we will cache the slices internally.
+impl Clone for FilterMask {
+    fn clone(&self) -> Self {
+        if self.selectivity > FILTER_SLICES_SELECTIVITY_THRESHOLD {
+            let _: VortexResult<_> = self
```

Comment:
so if there's an error we just... ignore it? where does this explode?

Reply:
ok, reading further down, it looks like if this fails, we'll reattempt it on every clone, but it will work bc it will fall back to LazySlices or LazyIndices. not sure that's the optimal behavior (at least the reattempts)

Reply:
🤷 we can't store a LazyLock<Result> since errors aren't clone able. Thing is, if any one of the cloned things fails to do this then the whole compute fails. It doesn't fallback to lazy iterator on failure, the error is propagated.

Reply:
this can only return an `Err` if `self.boolean_buffer()` returns one, in which case we're going to fail anyway